plugins/cookieauth.js

```diff
@@ -13,167 +13,90 @@
 // limitations under the License.
 'use strict';
 
-// this the the 'cookieauth' request handler.
-// Instead of passing the authentication credentials using HTTP Basic Auth with every request
-// we exchange the credentials for a cookie which we remember and pass back with each
-// subsequent request.
-var async = require('async');
-var debug = require('debug')('cloudant');
-var stream = require('stream');
-var u = require('url');
-var nullcallback = function() {};
+const debug = require('debug')('cookieauth');
+const request = require('request');
+const u = require('url');
 
-module.exports = function(options) {
-  var requestDefaults = options.requestDefaults || {};
-  var request = require('request').defaults(requestDefaults);
-  var jar = request.jar();
-  var cookieRefresh = null;
+const BasePlugin = require('./base.js');
 
-  // make a request using cookie authentication
-  // 1) if we have a cookie or have no credentials, just try the request
-  // 2) otherwise, get session cookie
-  // 3) then try the request
-  var cookieRequest = function(req, callback) {
-    // deal with absence of callback
-    if (typeof callback !== 'function') {
-      callback = nullcallback;
-    }
-
-    // parse the url to extract credentials and calculate
-    // stuburl - the cloudant url without credentials or auth
-    // auth - whether there are credentials or not
-    // credentials - object containing username & password
-    var url = req.uri || req.url;
-    var parsed = u.parse(url);
-    var auth = parsed.auth;
-    var credentials = null;
-    delete parsed.auth;
-    delete parsed.href;
-    url = u.format(parsed);
-    if (auth) {
-      var bits = auth.split(':');
-      credentials = {
-        username: bits[0],
-        password: bits[1]
-      };
-    }
-    req.url = url;
-    delete req.uri;
-    delete parsed.path;
-    delete parsed.pathname;
-    var stuburl = u.format(parsed).replace(/\/$/, '');
-
-    // to maintain streaming compatiblity, always return a PassThrough stream
-    var s = new stream.PassThrough();
-
-    // run these three things in series
-    async.series([
-
-      // call the request being asked for
-      function(done) {
-        // do we have cookie for this domain name?
-        var cookies = jar.getCookies(stuburl);
-        var statusCode = 500;
+function CookiePlugin(client, opts) {
+  CookiePlugin.super_.apply(this, arguments);
+  this.baseUrl = null;
+  this.cookieJar = request.jar();
+  this.credentials = null;
+  this.refreshRequired = true;
+}
+CookiePlugin.super_ = BasePlugin;
+CookiePlugin.prototype = Object.create(BasePlugin.prototype);
+CookiePlugin.prototype.onRequest = function(state, req, callback) {
+  req.jar = this.cookieJar;
```

Comment:
Can we check for the existence of the cookie and that it isn't expired already so we can preemptively request the `_session` and avoid the `401` round trip if we don't have one or it isn't valid. I think `refreshRequired=true` as the starting point handles the initial request, but if we can check the jar that would be good.

We will need this behaviour more for IAM sessions, but it is beneficial here too, probably it can be extracted out to some shared code (or added later with the IAM work) if that is preferable. Also it will need some tests to validate that expired cookies are handled by the built-in cookie jar or some purging logic will need adding.

Reply:
Something like this should remove expired cookies:
```js
If (this.baseUrl && req.jar.getCookies(this.baseUrl, {expire: true}).length > 0) {
  return callback(state);
}
```
Is the `refreshRequired` also needed? I was planning to set `refreshRequired = true` on 401/403. But perhaps just pruning the jar and calling `.length` on it is enough?

Reply:
I'm not sure how visible/cusotmizable the jar is, but as long as the thing we check for length is limited to our own `baseUrl` so we aren't counting cookies from other sources that may be in there then I think it would be OK.